Forward pass lightning - training pipeline

szmazurek opened the pull request:
Fixes #

## Proposed Changes
- full training logic implemented in the lightning module
- gandlf-native primitives transferred: saving, loading model, early stopping, checkpointing, logging
- currently, distributed setup is failing due to some problems with distributed sampler - this will probably need manual implementation


## Checklist


- [ ] [`CONTRIBUTING`](https://github.com/mlcommons/GaNDLF/blob/master/CONTRIBUTING.md) guide has been followed.
- [ ] PR is based on the [current GaNDLF master ](https://docs.github.com/en/desktop/contributing-and-collaborating-using-github-desktop/keeping-your-local-repository-in-sync-with-github/syncing-your-branch-in-github-desktop?platform=windows).
- [ ] Non-breaking change (does **not** break existing functionality): provide **as many** details as possible for _any_ breaking change.
- [ ] Function/class source code documentation added/updated (ensure `typing` is used to provide type hints, including and not limited to using `Optional` if a variable has a pre-defined value).
- [ ] Code has been [blacked](https://github.com/psf/black#usage) for style consistency and linting.
- [ ] If applicable, version information [has been updated in GANDLF/version.py](https://github.com/mlcommons/GaNDLF/blob/master/GANDLF/version.py).
- [ ] If adding a git submodule, add to list of exceptions for black styling in [pyproject.toml](https://github.com/mlcommons/GaNDLF/blob/master/pyproject.toml) file.
- [ ] [Usage documentation](https://github.com/mlcommons/GaNDLF/blob/master/docs) has been updated, if appropriate.
- [ ] Tests added or modified to [cover the changes](https://app.codecov.io/gh/mlcommons/GaNDLF); if coverage is reduced, please give explanation.
- [ ] If customized dependency installation is required (i.e., a separate `pip install` step is needed for PR to be functional), please ensure it is reflected in all the files that control the CI, namely: [python-test.yml](https://github.com/mlcommons/GaNDLF/blob/master/.github/workflows/python-test.yml), and all docker files [[1](https://github.com/mlcommons/GaNDLF/blob/master/Dockerfile-CPU),[2](https://github.com/mlcommons/GaNDLF/blob/devcontainer_build_fix/Dockerfile-CUDA11.6),[3](https://github.com/mlcommons/GaNDLF/blob/master/Dockerfile-ROCm)].
- [ ] The `logging` library is being used and no `print` statements are left.

szmazurek committed:
GANDLF/models/lightning_module.py

szmazurek committed:
Implementation of forward pass in lightning module

szmazurek committed:
Implementing info outputs

szmazurek committed:
Model checkpoint saving WIP

szmazurek committed:
Correct loss value retrieval for saving

szmazurek committed:
Early stoppoing & model saving

sarthakpati commented on GANDLF/models/lightning_module.py:
```diff
@@ -1,22 +1,53 @@
         super().__init__()
+        self.output_dir = output_dir
         self.params = deepcopy(params)
+        self.current_best_loss = 1e7
```
Does this make sense? 

```suggestion
        self.current_best_loss = sys.float_info.max
```

szmazurek replied:
totally, forgot about that!

sarthakpati commented on GANDLF/models/lightning_module.py:
```diff
@@ -34,6 +65,71 @@ def _initialize_preds_target_processor(self):
+                version_check(
+                    self.params["version"], version_to_check=checkpoint_dict["version"]
+                )
+                # I am purposefully ommiting the line below, as "previous_parameters" are not used anywhere
```
```suggestion
                # I am purposefully omitting the line below, as "previous_parameters" are not used anywhere
```

sarthakpati commented on GANDLF/models/lightning_module.py:
```diff
@@ -34,6 +65,71 @@ def _initialize_preds_target_processor(self):
+                self.trainer.callback_metrics["val_loss"] = checkpoint_dict["loss"]
+            except Exception as e:
+                warnings.warn(
+                    f"Previous best model found under path {self.model_paths['best']}, but error occured during loading: {e}; Continuing training with new model"
```
```suggestion
                    f"Previous best model found under path {self.model_paths['best']}, but error occurred during loading: {e}; Continuing training with new model"
```

sarthakpati commented on GANDLF/models/lightning_module.py:
```diff
@@ -42,4 +138,381 @@ def configure_optimizers(self):
+                len(self.params["value_keys"]),
+            )
+        else:
+            label = subject["label"][
+                torchio.DATA
+            ]  # segmentation; label is (B, C, H, W, D) image
```
```suggestion
            # segmentation; label is (B, C, H, W, D) image
            label = subject["label"][
                torchio.DATA
            ]  
```

sarthakpati commented on GANDLF/models/lightning_module.py:
```diff
@@ -42,4 +138,381 @@ def configure_optimizers(self):
+            labels = one_hot(labels, self.params["model"]["class_list"])
+
+        if self.params["model"]["dimension"] == 2:
+            images = images.squeeze(
+                -1
+            )  # removing depth, as torchio adds last dimension for 2D images
```
```suggestion
            # removing depth, as torchio adds last dimension for 2D images
            images = images.squeeze(
                -1
            )
```

sarthakpati commented on GANDLF/models/lightning_module.py:
```diff
@@ -42,4 +138,381 @@ def configure_optimizers(self):
+    def _set_epoch_start_time(self):
+        self.epoch_start_time = time.time()
+
+    # TODO when used with multple GPUs, thil will produce multiple logs
```
```suggestion
    # TODO when used with multiple GPUs, this should produce multiple logs
```

sarthakpati commented on GANDLF/models/lightning_module.py:
```diff
@@ -42,4 +138,381 @@ def configure_optimizers(self):
+        self.epoch_start_time = time.time()
+
+    # TODO when used with multple GPUs, thil will produce multiple logs
+    # for each GPU. We should think on doing allgather here in a function
```
```suggestion
    # for each GPU. We should think on doing altogether here in a function
```

szmazurek replied:
I actually really meant allgather operation for retrieving everything from metricas

sarthakpati replied:
Ah okay. Then maybe add that to the spell checker?

szmazurek replied:
Done

sarthakpati commented on GANDLF/models/lightning_module.py:
```diff
@@ -42,4 +138,381 @@ def configure_optimizers(self):
+        train_losses_gathered = self.all_gather(self.train_losses)
+        mean_loss = torch.mean(torch.stack(train_losses_gathered)).item()
+        # Note that we DO NOT have overall stats calculators here as in the original code
+        # Is it valid? Should we add it? This woudl require accumulation of all the predictions and labels
```
```suggestion
        # Is it valid? Should we add it? This would require accumulation of all the predictions and labels
```

sarthakpati commented on GANDLF/models/lightning_module.py:
```diff
@@ -42,4 +138,381 @@ def configure_optimizers(self):
+        train_losses_gathered = self.all_gather(self.train_losses)
+        mean_loss = torch.mean(torch.stack(train_losses_gathered)).item()
+        # Note that we DO NOT have overall stats calculators here as in the original code
+        # Is it valid? Should we add it? This woudl require accumulation of all the predictions and labels
+        # thus greatly increasing memory usage, especially in segmentation tasks
```
So, the `overall_stats` are only calculated for classification and regression tasks, and hence doesn't have any issues WRT memory usage. I feel we need to have this on.

szmazurek replied:
Right, somehow did not comprehend the conditional statement. Will update this one

szmazurek committed:
Update GANDLF/models/lightning_module.py

Co-authored-by: Sarthak Pati <patis@iu.edu>

szmazurek committed:
Update GANDLF/models/lightning_module.py

Co-authored-by: Sarthak Pati <patis@iu.edu>

szmazurek committed:
Update GANDLF/models/lightning_module.py

Co-authored-by: Sarthak Pati <patis@iu.edu>

szmazurek committed:
Update GANDLF/models/lightning_module.py

Co-authored-by: Sarthak Pati <patis@iu.edu>

szmazurek committed:
Update GANDLF/models/lightning_module.py

Co-authored-by: Sarthak Pati <patis@iu.edu>

szmazurek committed:
Update GANDLF/models/lightning_module.py

Co-authored-by: Sarthak Pati <patis@iu.edu>

szmazurek committed:
Update GANDLF/models/lightning_module.py

Co-authored-by: Sarthak Pati <patis@iu.edu>

szmazurek committed:
Training pass enhancements

- added calculation of overall metrics
- added proper handling of data formats for both progress bar logging
  and gandlf custom logging
- added tests for regression and classification
- little refactoring of tests

sarthakpati commented on .spelling/.spelling/expect.txt:
```diff
@@ -733,4 +733,6 @@ autograd
\ No newline at end of file
+Stdnet
+allgather
+skipif
```
```suggestion
skipif
progbar
```

sarthakpati committed:
Update .spelling/.spelling/expect.txt
